consensus/safety-rules/src/safety_rules_2chain.rs

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{error::Error, safety_rules::next_round, SafetyRules};
+use consensus_types::{
+    block::Block,
+    safety_data::SafetyData,
+    timeout_2chain::{TwoChainTimeout, TwoChainTimeoutCertificate},
+    vote::Vote,
+    vote_proposal::MaybeSignedVoteProposal,
+};
+use diem_crypto::{ed25519::Ed25519Signature, hash::CryptoHash, HashValue};
+use diem_types::{block_info::BlockInfo, ledger_info::LedgerInfo};
+
+/// 2-chain safety rules implementation
+impl SafetyRules {
+    pub(crate) fn guarded_sign_timeout_with_qc(
+        &mut self,
+        timeout: &TwoChainTimeout,
+        timeout_cert: Option<&TwoChainTimeoutCertificate>,
+    ) -> Result<Ed25519Signature, Error> {
+        self.signer()?;
+        let mut safety_data = self.persistent_storage.safety_data()?;
+        self.verify_epoch(timeout.epoch(), &safety_data)?;
+        self.verify_qc(timeout.quorum_cert())?;
+        if let Some(tc) = timeout_cert {
+            self.verify_tc(tc)?;
+        }
+
+        self.safe_to_timeout(timeout, timeout_cert, &safety_data)?;
+        if timeout.round() < safety_data.last_voted_round {
+            return Err(Error::IncorrectLastVotedRound(
+                timeout.round(),
+                safety_data.last_voted_round,
+            ));
+        }
+        if timeout.round() > safety_data.last_voted_round {
+            self.verify_and_update_last_vote_round(timeout.round(), &mut safety_data)?;
+            self.persistent_storage.set_safety_data(safety_data)?;
+        }
```

Comment:
It looks like a malicious caller can set a timeout.round() that is arbitrarily larger than safety_data.last_voted_round and advance the safety data round arbitrarily high (liveness attack).

Shouldn't only the TC, QC, or a vote be able to advance the safety data round?

I do note this can happen in the original sign_timeout() as well.

Reply:
advancing timeout round arbitrarily is actually a safety attack (with lower round qc), that's why we have safe_to_timeout check first

Reply:
Got it.  Not a part of this PR, but what about sign_timeout()?

Reply:
that's only about liveness and doesn't seem matter much

Reply:
I think it's not good from a TCB perspective that a corrupted validator can hurt liveness in the 3-chain protocol.  That being said, if we upgrade to 2-chain soon, perhaps this isn't as necessary to fix.